### PR TITLE
Ability to provide scopes from the request into the security backends

### PR DESCRIPTION
**Pull request by dbranco-sdc**
While we cleaned up the fork repository, we loose the delivered [pull request](https://github.com/greenpau/go-authcrunch/pull/16). This is just a cleaner pull request with exact same content for the new feature.

As mentioned [here](https://github.com/greenpau/caddy-security/issues/70#issue-1165485078), here's the first part of implementing the mentioned feature, so that if the AdditionalScopes flag is enabled in PolicyConfig, Authcrunch will forward the additional scopes provided by the client and append those to the ones already set on the identity provider.

We have on a `caddy-security` fork the second part ready, which is a flag on Caddyfile which allows for this functionality to be enabled, however we need a new release prior to doing this change since we cannot do a `replace` in the go.mod file to our fork 😀.

After this is merged and released we will do a pull request for the second part.

### CODE REVIEW

**Comment by greenpau** [comment]
@dbranco-sdc , for licensing purposes. Each person on the `Co-authored-by` must create its own PR with commit on `consent.yaml`. After they consented by creating the commits, I will accept this PR.

**Comment by dbranco-sdc** [comment]
@greenpau. Hello, I have seen that you have already added Paula and Sergiu consent into your branch repo.

**Comment by greenpau** [comment]
@dbranco-sdc , was not able to get to it over the weekend. Will try finding time on weekdays.

**Comment by greenpau** [comment]
@dbranco-sdc , I just finished with my current work on UI. Please rebase and re-test.

**Comment by dbranco-sdc** [comment]
> @dbranco-sdc , I just finished with my current work on UI. Please rebase and re-test.

It has been merge the latest changes from main into the pull request, and test still green

File: pkg/util/validate/additional_scopes.go [hunk 1]
```diff
@@ -0,0 +1,16 @@
+package validate
+
+import (
+	"github.com/greenpau/go-authcrunch/pkg/errors"
+	"regexp"
+)
+
+// AdditionalScopes verifies if the provided additional_scopes argument is valid
+func AdditionalScopes(additionalScopes string) error {
+	compile, _ := regexp.Compile("^[\\w|\\s]+$")
```
Comment (greenpau) [inline pkg/util/validate/additional_scopes.go:10]: @dbranco-sdc , please change to must compile and make it package-level.

File: pkg/authz/authenticate.go [hunk 2]
```diff
@@ -311,3 +315,14 @@ func (g *Gatekeeper) handleLoginHint(r *http.Request, ar *requests.Authorization
 		}
 	}
 }
+
+func (g *Gatekeeper) handleAdditionalScopes(r *http.Request, ar *requests.AuthorizationRequest) {
+	if additionalScopes := r.URL.Query().Get("additional_scopes"); additionalScopes != "" {
+		if err := validate.AdditionalScopes(additionalScopes); err != nil {
+			g.logger.Warn(err.Error())
```
Comment (greenpau) [inline pkg/authz/authenticate.go:322]: @dbranco-sdc , please see other logger message for reference. We would need more things that just the error.

File: pkg/authz/handlers/redirect.go [hunk 3]
```diff
@@ -124,12 +124,19 @@ func configureRedirect(w http.ResponseWriter, r *http.Request, rr *requests.Auth
...
 		loginHint := rr.Redirect.LoginHint
 		escapedLoginHint := url.QueryEscape(loginHint)
 		rr.Redirect.AuthURL = rr.Redirect.AuthURL + rr.Redirect.Separator + "login_hint" + "=" + escapedLoginHint
 		rr.Redirect.Separator = "&"
 	}
 
+	if len(rr.Redirect.AdditionalScopes) > 0 {
+		additionalScopes := rr.Redirect.AdditionalScopes
+		escapedAdditionalScopes := url.QueryEscape(additionalScopes)
+		rr.Redirect.AuthURL = rr.Redirect.AuthURL + rr.Redirect.Separator + "additional_scopes" + "=" + escapedAdditionalScopes
```
Comment (greenpau) [inline pkg/authz/handlers/redirect.go:137]: @dbranco-sdc , please use `fmt.Sprintf` for this.

File: pkg/util/validate/additional_scopes.go [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+	match := compile.MatchString(additionalScopes)
```
Comment (greenpau) [inline pkg/util/validate/additional_scopes.go:11]: @dbranco-sdc , could just be `if compile.MatchString ...`. Please rename `compile` to `pattern` or `r`.

**Review by greenpau — CHANGES_REQUESTED** [review]
@dbranco-sdc , please address my comments 👍

**Comment by dbranco-sdc** [comment]
> @dbranco-sdc , please address my comments 👍

Please check the latest changes implementing your suggestions, thanks in advance

**Comment by greenpau** [comment]
Thank you all for contributing! 👍

**Comment by greenpau** [comment]
@dbranco-sdc , please help with https://github.com/greenpau/caddy-security/issues/108
It might be related to the redirect changes.

**Pull request merged**